kimchi/src/circuits/domain_constant_evaluation.rs

```diff
@@ -0,0 +1,65 @@
+//! This contains the [DomainConstantEvaluations] which is used to provide precomputations to a [ConstraintSystem].
+
+use crate::circuits::domains::EvaluationDomains;
+use crate::circuits::polynomials::permutation::zk_polynomial;
+use crate::circuits::polynomials::permutation::ZK_ROWS;
+use ark_ff::FftField;
+use ark_poly::UVPolynomial;
+use ark_poly::{univariate::DensePolynomial as DP, Evaluations as E, Radix2EvaluationDomain as D};
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+
+use super::polynomials::permutation::vanishes_on_last_4_rows;
+
+#[serde_as]
+#[derive(Clone, Serialize, Deserialize, Debug)]
+/// pre-computed polynomials that depend only on the chosen field and domain
+pub struct DomainConstantEvaluations<F: FftField> {
+    /// 1-st Lagrange evaluated over domain.d8
+    #[serde_as(as = "o1_utils::serialization::SerdeAs")]
+    pub poly_x_d1: E<F, D<F>>,
+    /// 0-th Lagrange evaluated over domain.d4
+    // TODO(mimoo): be consistent with the paper/spec, call it L1 here or call it L0 there
+    #[serde_as(as = "o1_utils::serialization::SerdeAs")]
+    pub constant_1_d4: E<F, D<F>>,
+    /// 0-th Lagrange evaluated over domain.d8
+    #[serde_as(as = "o1_utils::serialization::SerdeAs")]
+    pub constant_1_d8: E<F, D<F>>,
+    /// the polynomial that vanishes on the last four rows
+    #[serde_as(as = "o1_utils::serialization::SerdeAs")]
+    pub vanishes_on_last_4_rows: E<F, D<F>>,
+    /// zero-knowledge polynomial over domain.d8
+    #[serde_as(as = "o1_utils::serialization::SerdeAs")]
+    pub zkpl: E<F, D<F>>,
+    #[serde_as(as = "o1_utils::serialization::SerdeAs")]
+    pub zkpm: DP<F>,
+}
+
+impl<F: FftField> DomainConstantEvaluations<F> {
+    pub fn create(domain: EvaluationDomains<F>) -> Option<Self> {
+        let poly_x_d1 = DP::from_coefficients_slice(&[F::zero(), F::one()])
```

Comment:
used to be l1

Reply:
btw this should be called `x_poly` maybe, the `d1` here is useless as we're not in an evaluations type